Automatic language selection + fix of Issue #16

nabbla opened the pull request:
Didn't have lot of time recently, also got stuck with making GraphicEx work on both VCL and FMX (but it will be done eventually), but have some good news, too.

I've wrote neat little lib called AutoResourceStr (going to document it and put to github soon), main purpose of which is exactly libs localization. With it programmer using a lib doesn't have to choose which GraphicStrings to add: he can add as many of them as he wants. When program is started, GraphicEx strings and error messages will be on correct language if available. Even better: if English-UK is set (or English-Australian etc), AutoResourceStr is smart enough to fetch English-US as the most appropriate. 

Good thing is: actual code is not changed, no need to replace simple resourcestring variable with something more complex (like t() in Qt which replaces original literal string with some translation), which was achieved using custom variant. It's rather old feature of Delphi (not some fancy generics or record operator overloading), so it works on delphi 7 as well as on all the newer ones.

Also, added 2 lines of code to GraphicEx to fix issue with paste from clipboard: we register clipboard formats again. Though, I couldn't reproduce that bug to begin with: no matter in what order I put Clipbrd and GraphicEx in uses clause, bitmap is pasted all right. Maybe it's not so in other versions, didn't yet try.

Commit:
Automatic language selection at runtime, using AutoResourceStr lib.

This way programmer using GraphicEx don't have to worry about choosing the right language by renaming proper GraphicStrings
and especially to manually put another ones into Delphi translation manager.

A little caveat so far: GraphicStrings should load before GraphicEx itself. AllGraphicStrings unit is added just for this purpose: you add this unit to Uses section in main form,
and normally everything will work correctly. If not all the languages are needed, then don't use AllGraphicStrings, but add the needed GraphicStringsXX to project, then in project file (menu Project -> Show source)
put them at first place to ensure that OpenPictureDialog filter is populated AFTER all the languages are loaded.

Commit:
Changed README.md, 'Compiling' section, brought it 'up to date'

Also, resolve of issue by tueddy (EClipboardException): added two lines TPicture.RegisterClipboardFormat()

mike-lischke commented on 3rd party/AutoResourceStr/variantWrapper.pas:
```diff
@@ -0,0 +1,249 @@
+procedure TAbstractWrapperVariantType.BinaryOp(var Left: TVarData; const Right: TVarData; const Operator: TVarOp);
+//var CastedRight: TVarData;
+begin
+//�������, �������, ��������, ��������, ������� �� ������� � ������� �������� (������, ���. � ��)
```
That looks a bit odd. Can you remove all the outcommented code and non-english comments? It's just adding noise.

mike-lischke commented on 3rd party/AutoResourceStr/variantWrapper.pas:
```diff
@@ -0,0 +1,249 @@
+
+function TAbstractWrapperVariantType.LeftPromotion(const V: TVarData; const Operator: TVarOp; out RequiredVarType: TVarType): Boolean;
+begin
+//�� ��� ������������� ���������� ����� �� �������� ��������, ����� �������� ����� �����������
```
same here

mike-lischke commented on 3rd party/AutoResourceStr/variantWrapper.pas:
```diff
@@ -0,0 +1,249 @@
+  else
+    RequiredVarType := VarType;
+  Result := True;
+//����� ��������� ������ ������, ����� "������������"
```
and here

mike-lischke commented:
That's pretty cool. Just clean the code up a bit and it's ready to be merged.

Commit:
- some formatting fixed and garbage of old comments and code
- AutoResourceStr variable may now be in expression between usual strings and behave normally (thanks to RightPromotion)
- works in Delphi 7 (needed language macros which were absent in windows.pas) and Delphi 10 Seattle (it uses VarUString variant type instead of VarString)
- should now work in x64 as well (before there was ambiguity about TAutoResourceStrVarData size in both cases)
- JPEGCompression now uses AutoResourceStr as well instead of string literals, its error messages are more generic (already translated to most languages)

Commit:
Added text file explaining what to do when GraphicStringsXX files can't be compiled in older versions of Delphi.

Commit:
README.md rewrote, now it's pretty optimistic, but hope the lib deserves it.

nabbla commented:
Ok, done that, checked under various versions of Delphi, it works, both x86 and x64.

There is some problem with non-unicode versions now: I've changed all string units except english to UTF-8 (otherwise all the extended characters may end up with wrong encoding), but now Delphi 7 refuses to open them because there are 3 characters at the beginning of file: EF BB BF. So to use them, they need to be reencoded to usual 1-byte chars.

I don't know is there many people using old versions, hope this change won't harm them too much. If they want multi-language application, they are doomed anyway, it was a real problem of wrong code pages, question marks everywhere etc., but the simple case of writing program for someone's native language was OK. 

Please look: I've changed README.md pretty much, maybe it's too optimistic now, don't know.

mike-lischke commented:
`EF BB BF` is the [Unicode BOM](https://en.wikipedia.org/wiki/Byte_order_mark) encoded in UTF-8. You can just open that file in a good text editor like Notepad++ and re-save it without BOM.

nabbla commented:
Good idea. I was afraid of deleting this BOM because thought: code editor won't recognize UTF-8 without it, but it did somehow.

Without this at least these files will compile normaly in Delphi 7, the only question remains: will it show extended characters properly when fed from UTF-8. Maybe AutoResourceStr itself could have a heuristic inside {$IFNDEF UNICODE} which detects UTF-8 at input and converts it to whatever the receiver side can understand.

mike-lischke commented:
Difficult to say. I don't remember if D7 is able to read UTF-8 encoded source files. On the other hand I believe the language pas files where all stored in UTF-8 all the time, so I don't think this will be a problem.

andersmelander commented:
Hi
We were just about to start the process of updating our local version of GraphicEx in all our projects when I discovered the changes made by this PR...

Did anybody consider that all existing localized projects would already be using another localization solution? The change from resourcestring to a homegrown system just means that 1) the strings that used to be localized will no longer be so and therefore 2) the new solution just adds unnecessary overhead. Nobody is going to base their localization on some auxiliary feature which happens to be in one of the libraries they use.

mike-lischke commented:
Hi Anders, this is the first complain about this patch I hear since more than 3 years. So I guess it either was fine for those people who were looking for a solution like this or almost nobody uses localization.

andersmelander commented:
Well if everything's fine then I guess we're just imagining things. Sorry to disturb.
